File: Week_03/51/n_queens.py
from typing import List
from collections import defaultdict


class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        row_pos = [0] * n
        column_used = defaultdict(bool)
        left_diagonal_used = defaultdict(bool)  # row + column
        right_diagonla_used = defaultdict(bool)  # row - column

        res = []

        def print_queue():
            template = [""] * n
            queue = []
            for i in range(n):
                for j in range(n):
                    if row_pos[i] == j:
                        template[j] = "Q"
                    else:
                        template[j] = "."
                queue.append("".join(template))
            return queue

        def dfs(row: int):
            if row == n:
                res.append(print_queue())
                return

            for col in range(n):
                if column_used[col] or left_diagonal_used[row + col] or right_diagonla_used[row - col]:
                    continue
                row_pos[row] = col
                column_used[col] = True
                left_diagonal_used[row + col] = True
                right_diagonla_used[row - col] = True
                dfs(row + 1)
                column_used[col] = False
                left_diagonal_used[row + col] = False
                right_diagonla_used[row - col] = False

        dfs(0)
        return res

File: Week_03/51/n_queens.go
package main

func solveNQueens(n int) [][]string {
	rowPos := make([]int, n)
	columnUsed := make([]bool, n)
	leftDiagonalUsed := make(map[int]bool, n*2)  // row + col
	rightDiagonalUsed := make(map[int]bool, n*2) // row - col
	res := make([][]string, 0)

	printQueen := func() []string {
		template := make([]byte, n)
		queen := make([]string, n)
		for i := 0; i < n; i++ {
			for j := 0; j < n; j++ {
				if rowPos[i] == j {
					template[j] = 'Q'
				} else {
					template[j] = '.'
				}
			}
			queen[i] = string(template)
		}
		return queen
	}
	var dfs func(row int)
	dfs = func(row int) {
		if row == n {
			res = append(res, printQueen())
			return
		}

		for column := 0; column < n; column++ {
			if columnUsed[column] || leftDiagonalUsed[row+column] || rightDiagonalUsed[row-column] {
				continue
			}
			rowPos[row] = column
			columnUsed[column] = true
			leftDiagonalUsed[row+column] = true
			rightDiagonalUsed[row-column] = true
			dfs(row + 1)
			columnUsed[column] = false
			leftDiagonalUsed[row+column] = false
			rightDiagonalUsed[row-column] = false
		}
	}
	dfs(0)
	return res
}
